### PR TITLE
Fix: Improve error message for empty device modes

### LINKED ISSUE #26629: Bad error message on empty invalid device mode in API requests
### Issue Description

While trying to use podman with the cluster-api Docker infrastructure provider the API request apparently contained a /dev/fuse:: device (I reported that as a bug here: https://github.com/kubernetes-sigs/cluster-api/issues/12485 ). This resulted in the unhelpful error message (both in the API response and the logs)
```
Request Failed(Internal Server Error): container create: invalid device mode: 
```
The same problem can be reproduced on the CLI with e.g.

```
# podman run -it --rm --device /dev/fuse:: debian:bookworm /bin/bash
Error: invalid device mode:
```
It would be useful in situations like this if the context (the entire device mapping string) was included in the error message. In the logs I would even suggest including more info from the request since most of the other information (e.g. the container name) only show up at significantly higher log levels.

### Steps to reproduce the issue

Steps to reproduce the issue
1. Request creation of container with device mapping /dev/fuse:: either from CLI or via API
2. observe error "invalid device mode: "

### Describe the results you received

invalid device mode:

### Describe the results you expected

Something along the lines of:

In device mapping "/dev/fuse::" for container "foo": device mappings do not allow empty device mode.
Either omit the second : or specify a valid device mode.

### podman info output

```yaml
host:
  arch: amd64
  buildahVersion: 1.40.1
  cgroupControllers:
  - cpuset
  - cpu
  - io
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: app-containers/conmon-2.1.13
    path: /usr/libexec/podman/conmon
    version: 'conmon version 2.1.13, commit: unknown'
  cpuUtilization:
    idlePercent: 94.27
    systemPercent: 1.57
    userPercent: 4.16
  cpus: 24
  databaseBackend: boltdb
  distribution:
    distribution: gentoo
    version: "2.17"
  eventLogger: journald
  freeLocks: 2010
  hostname: taladardesktop
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 6.15.6
  linkmode: dynamic
  logDriver: journald
  memFree: 7572860928
  memTotal: 67316932608
  networkBackend: netavark
  networkBackendInfo:
    backend: netavark
    dns:
      package: app-containers/aardvark-dns-1.15.0
      path: /usr/libexec/podman/aardvark-dns
      version: aardvark-dns 1.15.0
    package: app-containers/netavark-1.15.2
    path: /usr/libexec/podman/netavark
    version: netavark 1.15.2
  ociRuntime:
    name: crun
    package: app-containers/crun-1.21
    path: /usr/bin/crun
    version: |-
      crun version 1.21
      commit: 10269840aa07fb7e6b7e1acff6198692d8ff5c88
      rundir: /run/user/1000/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  pasta:
    executable: /usr/bin/pasta
    package: net-misc/passt-2025.06.11
    version: |
      pasta 2025.06.11
      Copyright Red Hat
      GNU General Public License, version 2 or later
        <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
      This is free software: you are free to change and redistribute it.
      There is NO WARRANTY, to the extent permitted by law.
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  rootlessNetworkCmd: pasta
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: app-containers/slirp4netns-1.2.0
    version: |-
      slirp4netns version 1.2.0
      commit: 656041d45cfca7a4176f6b7eed9e4fe6c11e8383
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.6.0
  swapFree: 0
  swapTotal: 0
  uptime: 61h 17m 4.00s (Approximately 2.54 days)
  variant: ""
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries: {}
store:
  configFile: /home/taladar/.config/containers/storage.conf
  containerStore:
    number: 1
    paused: 0
    running: 1
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/taladar/.local/share/containers/storage
  graphRootAllocated: 322055438336
  graphRootUsed: 284907409408
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Supports shifting: "false"
    Supports volatile: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 1874
  runRoot: /run/user/1000/containers
  transientStore: false
  volumePath: /home/taladar/.local/share/containers/storage/volumes
version:
  APIVersion: 5.5.2
  BuildOrigin: Gentoo Portage
  Built: 1752477684
  BuiltTime: Mon Jul 14 09:21:24 2025
  GitCommit: ""
  GoVersion: go1.24.5
  Os: linux
  OsArch: linux/amd64
  Version: 5.5.2
```

### Podman in a container

No

### Privileged Or Rootless

Rootless

### Upstream Latest Release

No

### Additional environment details

Gentoo Portage version of podman is also 5.5.2 just as latest upstream release but I rechnically did not try to download directly off GitHub releases and run it manually

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Devashish08**
- Add specific check for empty device modes in ParseDevice function
- Change error message from 'invalid device mode: ' to 'empty device mode in device specification: <device>'
- Include full device specification in error message for better context
- Add test cases for empty device mode scenarios
- Resolves issue where '/dev/fuse::' provided unhelpful error message

Fixes #26629

## Problem Description

When using device mappings with empty device modes (e.g., `/dev/fuse::`), users encounter an unhelpful error message that lacks context about which device specification caused the issue. This was reported in issue #26629 where cluster-api Docker infrastructure provider generated device mappings like `/dev/fuse::` resulting in the error: Error: invalid device mode:
The error message provides no context about which device specification is problematic, making debugging difficult in complex container configurations.

## User-Visible Impact

This affects users who:
- Use the Podman API with malformed device specifications  
- Use CLI commands with empty device modes like `podman run --device /dev/fuse::`
- Work with automated tools (like cluster-api) that may generate invalid device specifications

Without this fix, users cannot easily identify which device mapping is problematic, requiring manual inspection of all device specifications to find the issue.

## Technical Details

The fix adds specific empty mode detection in the `ParseDevice` function before calling `IsValidDeviceMode`. When an empty device mode is detected (e.g., in `/dev/fuse::`), it now returns a descriptive error message that includes the full device specification.

## Changes Made

- **pkg/specgen/generate/config_common.go**: Added empty device mode check in ParseDevice function
- **pkg/specgen/generate/config_common_test.go**: Added test case for empty device mode error  
- **test/e2e/run_device_test.go**: Added e2e test verifying the improved error message

## Testing

- ✅ Unit tests pass: `go test ./pkg/specgen/generate/ -v -run TestParseDevice`
- ✅ E2E tests pass with expected error message
- ✅ Valid device specifications continue to work correctly
- ✅ Other invalid modes still generate appropriate error messages

**Test command:**
```bash
./bin/podman run -it --rm --device /dev/fuse:: alpine:latest true
# Now outputs: Error: empty device mode in device specification: /dev/fuse::
```

Fixes: #26629

#### Does this PR introduce a user-facing change?

```release-note
Improved error message for empty device modes in device specifications. 
When using device mappings like `/dev/fuse::`, the error now clearly 
states "empty device mode in device specification" instead of the 
generic "invalid device mode" message, making it easier to identify 
and fix configuration issues.
```

### CODE REVIEW

**Comment by Devashish08** [comment]
@mheon any feedback on this PR, or whom to tag to review this PR

File: pkg/specgen/generate/config_common.go [hunk 1]
```diff
@@ -15,8 +15,11 @@ func ParseDevice(device string) (string, string, string, error) {
 	arr := strings.Split(device, ":")
 	switch len(arr) {
 	case 3:
+		if arr[2] == "" {
+			return "", "", "", fmt.Errorf("empty device mode in device specification: %s", device)
+		}
 		if !IsValidDeviceMode(arr[2]) {
-			return "", "", "", fmt.Errorf("invalid device mode: %s", arr[2])
+			return "", "", "", fmt.Errorf("invalid device mode: %s in device specification: %s", arr[2], device)
```
Comment (mheon) [inline pkg/specgen/generate/config_common.go]: "invalid device mode %q in device %q"

File: pkg/specgen/generate/config_common.go [hunk 2]
```diff
@@ -25,7 +28,7 @@ func ParseDevice(device string) (string, string, string, error) {
 			permissions = arr[1]
 		} else {
 			if len(arr[1]) > 0 && arr[1][0] != '/' {
-				return "", "", "", fmt.Errorf("invalid device mode: %s", arr[1])
+				return "", "", "", fmt.Errorf("invalid device mode: %s in device specification: %s", arr[1], device)
```
Comment (mheon) [inline pkg/specgen/generate/config_common.go]: Same as above

File: pkg/specgen/generate/config_common.go [hunk 3]
```diff
@@ -25,7 +28,7 @@ func ParseDevice(device string) (string, string, string, error) {
 			permissions = arr[1]
 		} else {
 			if len(arr[1]) > 0 && arr[1][0] != '/' {
-				return "", "", "", fmt.Errorf("invalid device mode: %s", arr[1])
+				return "", "", "", fmt.Errorf("invalid device mode %q in device %q", arr[1], device)
```
Comment (TomSweeneyRedHat) [inline pkg/specgen/generate/config_common.go:31]: Is this `/dev/fuse:`?  I'm not sure, but it would be nice to have a test for regardless.

Comment (Devashish08) [reply pkg/specgen/generate/config_common.go:31]: added test for it

**Comment by Devashish08** [comment]
@Luap99 can you review the PR

**Comment by Luap99** [comment]
Your unit tests are failing so please fix them and you also need to update other e2e test as they still expect the old error message, i.e.
https://api.cirrus-ci.com/v1/artifact/task/4578529767063552/html/int-podman-fedora-42-root-host-sqlite.log.html

**Comment by Devashish08** [comment]
@Luap99 can you tell how to generate this https://api.cirrus-ci.com/v1/artifact/task/4578529767063552/html/int-podman-fedora-42-root-host-sqlite.log.html
type of report to check if changes are working fine or not

**Comment by Luap99** [comment]
The html logs are generated by CI but if you mean to run the e2e test you need to use
```
make localintegration
```

Or because that takes a while and you don't need to run everything you can use
```
make localintegration FOCUS="<testname>"
```
https://github.com/containers/podman/tree/main/test#running-a-single-integration-test
(works on linux only)



But looks like your tests are green now, the remaining failures are flakes that you can ignore.

**Review by Luap99 — APPROVED** [review]
LGTM

**Comment by Devashish08** [comment]
@Luap99 @mheon is there anything else i need to do more in this PR or is this good to be merged?

**Comment by Luap99** [comment]
@Devashish08 This is good from my side, just waiting for a second review from someone else.
@mheon @TomSweeneyRedHat PTAL
